docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,181 @@
+# Kubelet and systemd interaction
+
+**Author**: Derek Carr (@derekwaynecarr)
+
+**Status**: Proposed
+
+This document presents a specification for how the `kubelet` interfaces with `systemd`.
+
+## Motivation
+
+Many Linux distributions have either adopted, or plan to adopt `systemd` as their init system.
+
+This document describes a series of enhancements that should be made to the `kubelet` to better integrate with these distributions.
+
+## systemd units
+
+To help understand this proposal, we first provide a brief summary of `systemd` behavior.
+
+`systemd` manages a hierarchy of `slice`, `scope`, and `service` units.
+
+* `service` - application on the server that is launched by `systemd`; how it should start/stop; when it should be started; under what 
+circumstances it should be restarted; and any resource controls that should be applied to it.
+* `scope` - a process or group of processes which are not launched by `systemd` (i.e. fork), like a service, resource controls may be applied
+* `slice` - organizes a hierarchy in which `scope` and `service` units are placed.  a `slice` may contain `slice`, `scope`, or `service` units;
+processes are attached to `service` and `scope` units only, not to `slices`. The hierarchy is intended to be unified, meaning a process should
+only belong to a single leaf node.
+
+## cgroup hierarchy: orthogonal versus unified hierarchies and cgroupfs single writer
+
+Classical `cgroup` hierarchies were orthogonal per resource group controller, and a process may exist in different parts of the hierarchy.
+
+For example, a process `p1` could exist in each of the following at the same time:
+
+* `/sys/fs/cgroup/cpu/important/`
+* `/sys/fs/cgroup/memory/unimportant/`
+* `/sys/fs/cgroup/cpuacct/unimportant/`
+
+In addition, controllers for one resource group could depend on another in ways that were not always obvious.
+
+For example, the `cpu` controller depends on the `cpuacct` controller yet they were treated separately.
+
+Many found it confusing for a single process to belong to different nodes in the `cgroup` hierarchy across controllers.
+
+The Kernel direction for `cgroup` support is to move toward a unified `cgroup` hierarchy, where the per-controller hierarchies are eliminated in
+favor of hierarchies like the following:
+
+* `/sys/fs/cgroup/important/`
+* `/sys/fs/cgroup/unimportant/`
+
+In a unified hierarchy, a process should only belong to a single node in the `cgroup` tree.
+
+The Kernel direction for `cgroup` management is to promote a single-writer model rather than allowing multiple processes to independently
+write to parts of the file-system.
+
+## cgroup management with systemd and container runtimes
+
+By default, `systemd` defines three slices:
+
+* `system.slice` - contains system and application daemons
+* `user.slice` - user login related processes
+* `machine.slice` - virtual machines or containers that have init systems (not application space containers)
```

Comment:
>  re machine.slice, I was under the impression it was not intended for **application sandbox containers**.  I thought Dan Walsh was looking at **running systemd itself in the container**

(emphasis mine) - my point is these two concepts are _not mutually exclusive_. rkt is for running application containers; it also (by default) uses systemd internally.


>  We have had issues in past with depth of libvirt and I will add the bugZilla discussion on that for context.

Please do. The rkt developers work very actively with cgroups, in upstream systemd and the kernel itself, and we'll definitely be interested in rooting out and solving any performance issues asap.

> Is there a pointer to what cgroup topology you have for pods on rocket today you can point me to?

Latest reference is [here](https://github.com/coreos/rkt/blob/master/Documentation/devel/cgroups.md). That document also links to some of the cgroups work we're doing, in rkt and aforementioned upstream projects.